### DISCUSSION TITLE
Unclamping vs. underPrior

### DISCUSSION THREAD

**Discussion by wrightaprilm** [revbayes/revbayes#623] [General]
Hi folks,

In our monthly meeting, we had a discussion of running things under the prior. We discussed two things that should be roughly equivalent, removing the clamps for all the data (discrete morphology and fossil ages in this case) and running with `underPrior = TRUE` in the `mcmc.run()` statement. 

But I get very different values for the prior with simply not clamping data and running explicitly `underPrior = TRUE`.  `underPrior = TRUE` undoes all the clamps, so not clamping anything should be equivalent, eh? If I echo the model to the screen, it doesn't look like any chunks of model are missing. So I'm not convinced I understand the behavior of `underPrior=TRUE` vs simply not clamping data to model objects. 

[diploploran_testCase.zip](https://github.com/user-attachments/files/17997500/diploploran_testCase.zip)

**Comment by davidcerny** [comment]
Your scripts don't seem to specify the random seed though – are you sure the differences are not just stochastic? What if you do specify it?

**Comment by wrightaprilm** [comment]
Good eye. I don't think that's the issue, though. Still seeing a difference of about 1200 units in the prior with the same seed.

[diploploran_testCase.zip](https://github.com/user-attachments/files/17997810/diploploran_testCase.zip)

**Comment by davidcerny** [comment]
Indeed... I think what's happening here is that our explanation of `underPrior=TRUE` as "it unclamps clamped nodes" was not 100% accurate. If you look at the source code, it doesn't actually call `.unclamp()`, which is a method of class `StochasticNode`, but `.setPriorOnly()`, which is a method of the base class `DagNode`.

**Comment by davidcerny** [comment]
Just copying this great explanation by @bredelings here:

> So... if you just comment out `y.clamp(...)`, then the probability for `y` should still be included.  If there are proposals for `y`, then you should effectively integrate it out.  But if there are no such moves, then it seems like the initial (simulated) value of `y`  could  constrain the analysis?
> 
> If, on the other hand, you set the probability for `y` to `1.0` , then it shouldn't have an effect on the analysis.
> 
> And you wouldn't spend time computing the probability of `y`.

**Comment by wrightaprilm** [comment]
The discrepancy between what I expected and the behavior, while it makes sense, makes me more in favor of a solution as you proposed:

`phySeq ~ dnPhyloCTMC(tree=timetree, Q=Q_cytb, siteRates=rates_cytb, branchRates=branch_rates, type="DNA")
phySeq.clamp(cytb)
phySeq.setPriorOnly()`
